File: arrayproductexceptself.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        #initializing result array
        result=[0 for i in range(len(nums))]
        rp=1
        temp=1
        #product of left elements to element
        for i in range(0,len(nums)):
            rp=rp*temp
            result[i]=rp
            temp=nums[i]
        temp=1
        rp=1
        #product of left elements and right elements to element
        for i in range(len(nums)-1,-1,-1):
            rp=rp*temp
            result[i]*=rp
            temp=nums[i]
        return result
#Time Complexity: O(n)
#Space Complexity: O(1) excluding result 

File: ArrayProductExceptSelf.java
//time Complexity: O(N)
//Space Complexity O(1) Optimized part , for basic we can use left and right separate array which will take O(n) space as well
//Leetcode: yes.
public class ArrayProductExceptSelf {
        public int[] productExceptSelf(int[] nums) {

            if (nums == null || nums.length == 0) {
                return new int[] {};
            }

            int len = nums.length;
            int[] result = new int[len];

            result[0] = 1;
            //calculate the product of all the numbers to the left of the current index.
            for (int i = 1; i < len; i++) {
                result[i] = result[i-1] * nums[i-1];
            }
            //calculate the product of all the numbers to the right of the current index
            //and then combine it with the previous calculated product to give us left * right
            int r = 1;
            for (int i = len-1; i >= 0; i--) {
                result[i] = result[i] * r;
                r  = r * nums[i];
            }

            return result;
        }
}
